### DISCUSSION TITLE
Python does not seem to understand default values in dataclasses

### DISCUSSION THREAD

**Discussion by Makistos** [sqlalchemy/sqlalchemy#10476] [Usage Questions] [answered]
I have this sort of inherited class:
```
class Trackable(CommonMixin):
    ''' Mixin for tables that have id, created_at and updated_at columns.
    '''
    id: Mapped[int] = mapped_column(primary_key=True)
    created_at: Mapped[Optional[datetime.datetime]] = mapped_column(
        DateTime, server_default=func.now())
    updated_at: Mapped[Optional[datetime.datetime]] = mapped_column(
        DateTime, onupdate=func.now())

    def _serialize(self) -> dict[str, Any]:
        return {
            'id': self.id,
            'created_at': self.created_at,
            'updated_at': self.updated_at
        }
```
Which I use in a class like this:
```
class cost_centers(Trackable, Base):
    name: Mapped[str] = mapped_column(String(255))

    def serialize(self) -> dict[str, Any]:
        return {**self._serialize(),
                **{
            'name': self.name
        }}
```
But if I try to create a cost_centers item Python insists I need to give id, created_at and updated_at fields as well. If I add default values, app won't even start because default values are defined before other fields in the inherited classes. So how am I supposed to use server_default and onupdate fields?

**Comment by CaselIT** [comment, marked as answer]
Hi,

If you use dataclasses then the normal dataclasses rules apply. You can set `init=False` on the field you don't want to set in the init. Another alternative is to use `init=False` in the class definition so that the default sqlalchemy init can ve used

**Reply by Makistos** [reply]
That seems to work for updated_at field, but I still get this error:

```
ERROR in cost_centers: create_cost_center: (pymysql.err.OperationalError) (1364, "Field 'created_at' doesn't have a default value")
[SQL: INSERT INTO cost_centers (name, updated_at) VALUES (%(name)s, %(updated_at)s) RETURNING cost_centers.id, cost_centers.created_at]
[parameters: {'name': '100', 'updated_at': None}]
```
I only give the name parameter when creating the object so this is rather strange.

**Reply by CaselIT** [reply]
That seems that the sql table has been defined without a server default.
Ensure that the correct definition is used to create the table

**Reply by Makistos** [reply]
Ah, I see. This database was given to me and I have just added some missing constraints to it. I think it's in this case easier to manually set the initial values than alter every table that uses those fields.

**Reply by CaselIT** [reply]
you can tell sqlalchemy to use now() when inserting by using https://docs.sqlalchemy.org/en/20/orm/mapping_api.html#sqlalchemy.orm.mapped_column.params.insert_default

or you can use alembic or similar to add a default to the sql table using a migration

**Reply by Makistos** [reply]
I ended up altering the tables anyways, this was the easiest and probably most foolproof method.
